File: packages/grammar/src/ts/error-listener.ts
// Error listener for collecting parsing errors
import type { ANTLRErrorListener, RecognitionException, Recognizer } from "antlr4ts";
import type { ParseError } from "@aa/types";

export class CollectingErrorListener implements ANTLRErrorListener<any> {
  public errors: ParseError[] = [];

  syntaxError(
    recognizer: Recognizer<any, any>,
    offendingSymbol: any,
    line: number,
    charPositionInLine: number,
    msg: string,
    e: RecognitionException | undefined
  ): void {
    this.errors.push({
      line,
      column: charPositionInLine,
      message: msg,
    });
  }
}

File: packages/grammar/py/src/aa_grammar/error_listener.py
from antlr4.error.ErrorListener import ErrorListener  # type: ignore
from typing import List, Dict, Any

class CollectingErrorListener(ErrorListener):
    def __init__(self) -> None:
        super().__init__()
        self.errors: List[Dict[str, Any]] = []

    def syntaxError(self, recognizer, offendingSymbol, line, column, msg, e):
        # ANTLR: line base 1, column base 0 → normalizar a base 1
        self.errors.append({
            "line": int(line),
            "column": int(column) + 1,
            "message": str(msg),
        })
